Mathlib/Probability/Kernel/IonescuTulcea/Traj.lean

```diff
@@ -0,0 +1,757 @@
+`Kernel (Π i : Iic a, X i) (Π n ≥ a, X n)` which given the trajectory up to time `a` outputs
+the distribution of the infinite trajectory started in `X (a + 1)`. In other words this theorem
+makes sense of composing infinitely many kernels together.
+
+In this file we construct this "limit" kernel given the family `κ`. More precisely, for any `a : ℕ`,
+we construct the kernel `traj κ a : Kernel (Π i : Iic a, X i) (Π n, X n)`, which takes as input
+the trajectory in `X 0 × ... × X a` and outputs the distribution of the whole trajectory. The name
+`traj` thus stands for "trajectory". We build a kernel with output in `Π n, X n` instead of
+`Π i ≥ a, X i` to make manipulations easier. The first coordinates are deterministic.
+
+We also provide to compute integrals against `traj κ a` and an expression for the conditional
+expectation.
+
+## Main definition
+
+* `traj κ a`: a kernel from `Π i : Iic a, X i` to `Π n, X n` which takes as input a trajectory
+  up to time `a` ad outputs the distribution of the trajectory obtained by iterating the kernels
+  `κ`. Its existence is given by the Ionescu-Tulcea theorem.
+
+## Main statements
+
+* `eq_traj`: Uniqueness of `traj`: to check that `η = traj κ a` it is enough to show that
+  the restriction of `η` to variables `≤ b` is `partialTraj κ a b`.
+* `traj_comp_partialTraj`: Given the distribution up to tome `a`, `partialTraj κ a b`
+  gives the distribution of the trajectory up to time `b`, and composing this with
+  `traj κ b` gives the distribution of the whole trajectory.
+* `condExp_traj`: If `a ≤ b`, the conditional expectation of `f` with respect to `traj κ a`
+  given the information up to time `b` is obtained by integrating `f` against `traj κ b`.
+
+## Implementation notes
+
+The kernel `traj κ a` is built using the Carathéodory extension theorem. First we build a projective
+family of measures using `inducedFamily` and `partialTraj κ a`. Then we build an
+`MeasureTheory.AddContent` on `MeasureTheory.measurableCylinders` called `trajContent` using
+`projectiveFamilyContent`. Finally we prove `trajContent_tendsto_zero` which implies the
+`σ`-additivity of the content, allowing to turn it into a measure.
+
+## References
+
+We follow the proof of Theorem 8.24 in
+[O. Kallenberg, *Foundations of Modern Probability*][kallenberg2021]. For a more detailed proof
+in the case of constant kernels (i.e. measures),
+see Proposition 10.6.1 in [D. L. Cohn, *Measure Theory*][cohn2013measure].
+
+## Tags
+
+Ionescu-Tulcea theorem
+-/
+
+open Filter Finset Function MeasurableEquiv MeasurableSpace MeasureTheory Preorder ProbabilityTheory
+
+open scoped ENNReal Topology
+
+variable {X : ℕ → Type*}
+
+/-- Gluing `Iic a` and `Ioi a` into `ℕ`, version as a measurable equivalence
+on dependent functions. -/
+def MeasurableEquiv.IicProdIoi [∀ n, MeasurableSpace (X n)] (a : ℕ) :
+    ((Π i : Iic a, X i) × ((i : Set.Ioi a) → X i)) ≃ᵐ (Π n, X n) where
```

Comment:
You use two different notations for product spaces on the same line. Let's use always the same one.
```suggestion
    ((Π i : Iic a, X i) × (Π i : Set.Ioi a, X i)) ≃ᵐ Π n, X n where
```